### PR TITLE
NFDIV-4012: Add and update change of representative tab.

### PR DESCRIPTION
**Pull request by FaisalMoJ**
### Change description ###

Add and update change of representative tab.

### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/NFDIV-4012

### Pull request checklist ###

**Before raising**
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] README and other documentation has been updated / added (if needed)

**Before merging**
- [ ] this ticket been reviewed by QA
- [ ] the user story been signed off by the PO

**Note:** Bug fixes, dependency updates and technical tasks do not directly impact the user experience and can be merged without QA and PO review.

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/divorce/noticeofchange/event/SystemApplyNoticeOfChange.java [hunk 1]
```diff
@@ -100,31 +109,54 @@ private void updateChangeOfRepresentation(CaseDetails<CaseData, State> details,
...
-        if (APPLICANT_1_SOLICITOR.getRole().equals(changeOrganisationRequest.getCaseRoleId().getRole())) {
+        if (isApplicant1) {
+            party = "Applicant";
+            addedRepresentative = updateRepresentative(updatedBy, loggedInUserEmail, orgToAdd);
+            removedRepresentative = updateRepresentative(applicant1Solicitor.getName(),
+             applicant1Solicitor.getEmail(), applicant1Solicitor.getOrganisationPolicy().getOrganisation());
             updateOrgPolicyAndSolicitorDetails(applicant1Solicitor,
                     nocRequestingUser, loggedInUserEmail);
         } else {
+            party = SOLE_APPLICATION.equals(caseData.getApplicationType()) ? "Respondent" : "Applicant2";
```
[the remark this answers is not in the document]
Comment (FaisalMoJ) [reply src/main/java/uk/gov/hmcts/divorce/noticeofchange/event/SystemApplyNoticeOfChange.java:138]: Discussed with Adam, we will add an AC for this to be reproduced and then see if we need to do something about it make it more clear when switch to sole event occurred on the case

File: src/main/java/uk/gov/hmcts/divorce/noticeofchange/service/ChangeOfRepresentativeService.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+                .map(FindUsersByOrganisationResponse::getUsers)
+                .orElse(Collections.emptyList());
+
+        return organisationUsers.stream()
+                .filter(user -> user.getEmail().equalsIgnoreCase(loggedInUserEmail))
+                .findFirst()
+                .orElseThrow();
+    }
+
+    private Representative updateRepresentative(String solicitorName, String solicitorEmail,
```
Comment (BWeeks-JusticeGovUK) [inline src/main/java/uk/gov/hmcts/divorce/noticeofchange/service/ChangeOfRepresentativeService.java:113]: Maybe simplify this to accept a single parameter consisting of the solicitor object, then get the fields needed within this method itself?

File: src/main/java/uk/gov/hmcts/divorce/noticeofchange/service/ChangeOfRepresentativeService.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        var currentSolicitor = isApplicant1 ? caseData.getApplicant1().getSolicitor()  : caseData.getApplicant2().getSolicitor();
+
+        if (SOLICITOR_NOTICE_OF_CHANGE.getValue().equals(updatedVia)) {
+            String sysUserToken = idamService.retrieveSystemUpdateUserDetails().getAuthToken();
+            String s2sToken = authTokenGenerator.generate();
+            var changeOrganisationRequest = caseData.getChangeOrganisationRequestField();
+            var orgToAdd = changeOrganisationRequest.getOrganisationToAdd();
+            var organisationId = orgToAdd.getOrganisationId();
+            var loggedInUserEmail = changeOrganisationRequest.getCreatedBy().toLowerCase();
+            final var nocRequestingUser = getProfessionalUsers(sysUserToken, s2sToken, organisationId, loggedInUserEmail);
```
Comment (BWeeks-JusticeGovUK) [inline src/main/java/uk/gov/hmcts/divorce/noticeofchange/service/ChangeOfRepresentativeService.java:56]: Lots of var usage - could we specify types?  Might be overkill.

**Review by BWeeks-JusticeGovUK — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/divorce/noticeofchange/service/ChangeOfRepresentativeService.java (hunk 2 above)
Comment (FaisalMoJ) [reply src/main/java/uk/gov/hmcts/divorce/noticeofchange/service/ChangeOfRepresentativeService.java:109]: Added a method to accept single paramater

File: src/main/java/uk/gov/hmcts/divorce/noticeofchange/event/SystemApplyNoticeOfChange.java [hunk 4]
```diff
@@ -100,31 +109,54 @@ private void updateChangeOfRepresentation(CaseDetails<CaseData, State> details,
...
+        String party;
+        Representative addedRepresentative;
+        Representative removedRepresentative;
 
-        if (APPLICANT_1_SOLICITOR.getRole().equals(changeOrganisationRequest.getCaseRoleId().getRole())) {
+        if (isApplicant1) {
+            party = "Applicant";
+            addedRepresentative = updateRepresentative(updatedBy, loggedInUserEmail, orgToAdd);
+            removedRepresentative = updateRepresentative(applicant1Solicitor.getName(),
+             applicant1Solicitor.getEmail(), applicant1Solicitor.getOrganisationPolicy().getOrganisation());
```
[the remark this answers is not in the document]
Comment (FaisalMoJ) [reply src/main/java/uk/gov/hmcts/divorce/noticeofchange/event/SystemApplyNoticeOfChange.java]: Refactored the code to introduce a new class with all the logic about tab

File: src/test/java/uk/gov/hmcts/divorce/noticeofchange/event/SystemApplyNoticeOfChangeTest.java [hunk 5]
```diff
@@ -155,9 +169,18 @@ void shouldApplyNoticeOfChangeForApplicant2Solicitor() {
                 TEST_AUTHORIZATION_TOKEN, TEST_SERVICE_AUTH_TOKEN, acaRequest
         );
 
+        var changeOfRepresentative = applicant2CaseData.getChangeOfRepresentatives().stream()
```
[the remark this answers is not in the document]
Comment (FaisalMoJ) [reply src/test/java/uk/gov/hmcts/divorce/noticeofchange/event/SystemApplyNoticeOfChangeTest.java]: Refactored this class and tab building from citizen -> sol is covered in caseworker notice of change. If this is still required i will add a test

File: src/main/java/uk/gov/hmcts/divorce/noticeofchange/service/ChangeOfRepresentativeService.java (hunk 3 above)
Comment (FaisalMoJ) [reply src/main/java/uk/gov/hmcts/divorce/noticeofchange/service/ChangeOfRepresentativeService.java]: Replaced var with actual types

**Review by adamg-hmcts — APPROVED** [review]

**Pull request merged**